net/netns/netns_darwin.go

```diff
@@ -34,15 +42,109 @@ func controlLogf(logf logger.Logf, network, address string, c syscall.RawConn) e
 		// Don't bind to an interface for localhost connections.
 		return nil
 	}
-	idx, err := interfaces.DefaultRouteInterfaceIndex()
+
+	addr, err := netip.ParseAddr(address)
 	if err != nil {
-		logf("[unexpected] netns: DefaultRouteInterfaceIndex: %v", err)
+		logf("[unexpected] netns: error parsing address %q: %v", address, err)
 		return nil
 	}
 
+	var idx int
+	if bindToDeviceByRoute() {
+		idx, err = interfaceIndexFor(addr)
+
+		// Verify that we didn't just choose the Tailscale interface;
+		// if so, we fall back to binding from the default.
+		if err == nil {
+			_, tsif, err2 := interfaces.Tailscale()
+			if err2 == nil && tsif.Index == idx {
+				err = errInterfaceIndexInvalid
+			}
+		}
+	} else {
+		err = errInterfaceIndexInvalid
+	}
+	if err != nil {
+		idx, err = interfaces.DefaultRouteInterfaceIndex()
+		if err != nil {
+			logf("[unexpected] netns: DefaultRouteInterfaceIndex: %v", err)
+			return nil
+		}
+	}
+
 	return bindConnToInterface(c, network, address, idx, logf)
 }
 
+// interfaceIndexFor returns the interface index that we should bind to in
+// order to send traffic to the provided address.
+func interfaceIndexFor(addr netip.Addr) (int, error) {
```

Comment:
Depending on how expensive this is, we could add a cache for this (and invalidate it via ` LinkMonitor` `RegisterChangeCallback` callback).

Reply:
This is pretty expensive, yeah, which is why I'm not planning on enabling this by default. In the long term, I want to move ~all of this to the `LinkMonitor` by having it maintain a copy of the system's route table so we don't need to re-parse everything.

In the mean time, happy to do this; any suggestions or thoughts on where the callback should be registered, since we don't really have persistent state conveniently available?

Reply:
I think for now we don't have to do any of it, was just speculating.

When we do start to do it, triggering the cache invalidation from `LocalBackend`'s `linkChange` seems like a reasonable place (as far as `LocalBackend` also enabling this code path to start with).